Add computation of size contribution to info verb

nicolaloi opened the pull request:
Addresses #1601 feature request (Show size contribution of each topic with ros2 bag info). This is a test draft.

Output of `ros2 bag info test_bag -v --size-contribution`:
```shell
Files:             test_bag.mcap
Bag size:          2.6 GiB
Storage id:        mcap
ROS Distro:        unknown
Duration:          39.541s
Start:             Jun 19 2024 23:40:38.345 (1718833238.345)
End:               Jun 19 2024 23:41:17.886 (1718833277.886)
Messages:          2310
Topic information: Topic: /parameter_events | Type: rcl_interfaces/msg/ParameterEvent | Count: 0 | Serialization Format: cdr
                   Topic: /velodyne_points | Type: sensor_msgs/msg/PointCloud2 | Count: 765 | Size Contribution: 557.0 MiB | Serialization Format: cdr
                   Topic: /events/read_split | Type: rosbag2_interfaces/msg/ReadSplitEvent | Count: 0 | Serialization Format: cdr
                   Topic: /image_raw_1 | Type: sensor_msgs/msg/Image | Count: 763 | Size Contribution: 1.0 GiB | Serialization Format: cdr
                   Topic: /image_raw_2 | Type: sensor_msgs/msg/Image | Count: 763 | Size Contribution: 1.0 GiB | Serialization Format: cdr
                   Topic: /events/write_split | Type: rosbag2_interfaces/msg/WriteSplitEvent | Count: 0 | Serialization Format: cdr
                   Topic: /rosout | Type: rcl_interfaces/msg/Log | Count: 19 | Size Contribution: 4.9 KiB | Serialization Format: cdr
Service:           0
Service information:
```

The total size contribution of each topic is computed by sequentially reading every message in the rosbag. As mentioned in #1601, this approach can be slow. 

In a first test A (previous output), the computation of the size contribution took ~520ms for a bag with 2310 messages, 2.6GiB large, 39.5s long, with two _sensor_msgs/msg/Image_ topics with 763 messages each, and one _sensor_msgs/msg/PointCloud2_ topic with 765 messages.

In a second test B, the size computation took ~78ms for a bag with 14577 messages, 5.2MiB large, 74.4s long, with one _sensor_msgs/msg/Imu_ topic with 14559 messages. 

So to me, it seems the computation time is not simply dependent on the number of messages, but on their types and individual sizes too. As a test, I was curious about skipping the message serialization step inside the reader. I tried a brute modification of the mcap_storage interface to directly access the [messageView.message.dataSize](https://github.com/ros2/rosbag2/blob/fa84cfeeb0b640d53c23d25634ec5b0177ae1efe/rosbag2_storage_mcap/src/mcap_storage.cpp#L560C67-L560C95) variable while skipping the message serialization step. However, this only resulted in a 15-30% improvement in computation time. 

As a test, removing the actual size computation code and just reading the rosbag in an empty while loop doesn't really change the timing (only real difference was in test B, 78ms->68ms).

In `rosbag2_py/_info.cpp`, the new function [`compute_topics_size_contribution`](https://github.com/ros2/rosbag2/blob/a67f677fc93693faec0a45524b8fa333b3844056/rosbag2_cpp/src/rosbag2_cpp/info.cpp#L168) can be combined with the existing [`read_service_info`](https://github.com/ros2/rosbag2/blob/a67f677fc93693faec0a45524b8fa333b3844056/rosbag2_cpp/src/rosbag2_cpp/info.cpp#L72) to avoid reading the rosbag twice.

@MichaelOrlov looking forward to your feedback and suggested changes.

nicolaloi committed:
Add optional computation of size contribution to info verb

Signed-off-by: Nicola Loi <nicolaloi@outlook.com>

MichaelOrlov commented on ros2bag/ros2bag/verb/info.py:
```diff
@@ -50,8 +54,11 @@ def main(self, *, args):  # noqa: D102
             print("Warning! You have set both the '-t' and '-v' parameters. The '-t' parameter "
                   'will be ignored.')
+        if not args.verbose and args.size_contribution:
+            print("Warning! You have set the '--size-contribution' parameter without the '-v' parameter. "
+                  "The '--size-contribution' parameter will be ignored.")
         if args.verbose:
-            Info().read_metadata_and_output_service_verbose(args.bag_path, args.storage)
+            Info().read_metadata_and_output_service_verbose(args.bag_path, args.storage, args.size_contribution)
```
I don't think that we need extra parameter like `--size-contribution` in addition to the `--verbose` argument.
Let's just add size calculation by default to the `--verbose` CLI option.

MichaelOrlov commented on rosbag2_cpp/src/rosbag2_cpp/info.cpp:
```diff
@@ -165,4 +165,27 @@ std::vector<std::shared_ptr<rosbag2_service_info_t>> Info::read_service_info(
+  rosbag2_storage::ReadOrder read_order;
+  if (!storage->set_read_order(read_order)) {
+    throw std::runtime_error("Failed to set read order on " + uri);
+  }
```
```suggestion
```

We don't know and don't really care what the read order will be.

MichaelOrlov commented on ros2bag/ros2bag/verb/info.py:
```diff
@@ -50,8 +54,11 @@ def main(self, *, args):  # noqa: D102
                   'will be ignored.')
+        if not args.verbose and args.size_contribution:
+            print("Warning! You have set the '--size-contribution' parameter without the '-v' parameter. "
+                  "The '--size-contribution' parameter will be ignored.")
```
I don't think that we need extra parameter like `--size-contribution` in addition to the `--verbose` argument.
Let's just add size calculation by default to the `--verbose` CLI option.

MichaelOrlov commented on ros2bag/ros2bag/verb/info.py:
```diff
@@ -50,8 +54,11 @@ def main(self, *, args):  # noqa: D102
+            print("Warning! You have set the '--size-contribution' parameter without the '-v' parameter. "
+                  "The '--size-contribution' parameter will be ignored.")
         if args.verbose:
-            Info().read_metadata_and_output_service_verbose(args.bag_path, args.storage)
+            Info().read_metadata_and_output_service_verbose(args.bag_path, args.storage, args.size_contribution)
         else:
             m = Info().read_metadata(args.bag_path, args.storage)
             if args.topic_name:
```
We need to redesign a bit of a set of functions. 
For the sake of clarity, we need to have 3 functions:
1. Info().read_metadata
2. Info().print_output(metadata)
3. Info().print_output_verbose(metadata). i.e., instead of the `Info().read_metadata_and_output_service_verbose(..)`

nicolaloi replied:
Done, and also added Info().print_output_topic_name_only(metadata)

MichaelOrlov commented on rosbag2_py/src/rosbag2_py/format_bag_metadata.cpp:
```diff
@@ -118,10 +119,14 @@ void format_topics_with_type(
   }
 
   auto print_topic_info =
-    [&info_stream](const rosbag2_storage::TopicInformation & ti) -> void {
+    [&info_stream, &topics_size](const rosbag2_storage::TopicInformation & ti) -> void {
       info_stream << "Topic: " << ti.topic_metadata.name << " | ";
       info_stream << "Type: " << ti.topic_metadata.type << " | ";
       info_stream << "Count: " << ti.message_count << " | ";
+      auto topic_size_iter = topics_size.find(ti.topic_metadata.name);
+      if (topic_size_iter != topics_size.end()) {
+        info_stream << "Size Contribution: " << format_file_size(topic_size_iter->second) << " | ";
+      }
       info_stream << "Serialization Format: " << ti.topic_metadata.serialization_format;
       info_stream << std::endl;
     };
```
It would be great to have consistent output about `Size contribution: ` and print it out even for topics with zero messages and zero size contribution. 
We can distinguish verbose/no-verbose cases by checking `topics_size.size()` or, even better, by adding an extra `bool verbose` parameter, to the function for future extensions. Note: we shall not print `Size contribution: ` in non-verbose mode.

The rationale for the consistent output is that some external script that could parse output from the `ros2 bag info --verbose` can rely on the `Size contribution: ` field to be present for each topic.

nicolaloi replied:
Done, and also added size contribution for services.

MichaelOrlov reviewed (commented):
@nicolaloi Thanks for your contribution. 
Without deserialization, the message's size calculation per topic will be pretty fast in the vast majority of cases. Perhaps no more than a few seconds. I think adding a separate CLI argument like `--size-contribution` doesn't make sense. It will be much simpler to make output for topics size contribution by default for the `--verbose` CLI option.

See my detailed comments and suggestions inline.

nicolaloi committed:
Update rosbag2_cpp/src/rosbag2_cpp/info.cpp

Co-authored-by: Michael Orlov <morlovmr@gmail.com>
Signed-off-by: Nicola Loi <nicolaloi@outlook.com>

nicolaloi committed:
Fixes for review and failed tests

- Also update rosbag2_tests

Signed-off-by: Nicola Loi <nicolaloi@outlook.com>

nicolaloi committed:
Support services' size

- Also add new test and update design doc

Signed-off-by: Nicola Loi <nicolaloi@outlook.com>

nicolaloi commented:
Addressed review comments, added size contribution for services, and updated tests and design doc.

MichaelOrlov commented:
@ros-pull-request-builder retest this please

nicolaloi commented:
I will edit the file with the style divergence error. I have missed it since locally the tests no longer gave me any errors.

nicolaloi committed:
Fix style divergence

Signed-off-by: Nicola Loi <nicolaloi@outlook.com>

MichaelOrlov commented on rosbag2_py/src/rosbag2_py/_info.cpp:
```diff
@@ -41,12 +42,30 @@ class Info
-  void read_metadata_and_output_service_verbose(
-    const std::string & uri,
-    const std::string & storage_id)
+  void print_output(
+    rosbag2_storage::BagMetadata metadata_info)
```
Nitpick: Please use const reference and it could be in one line
```suggestion
 void print_output(const rosbag2_storage::BagMetadata & metadata_info)
```

MichaelOrlov commented on rosbag2_py/src/rosbag2_py/_info.cpp:
```diff
@@ -41,12 +42,30 @@ class Info
+    std::cout << format_bag_meta_data(metadata_info) << std::endl;
+  }
+
+  void print_output_topic_name_only(
+    rosbag2_storage::BagMetadata metadata_info)
```
Nitpick: Please use const reference and it could be in one line
```suggestion
  void print_output_topic_name_only(const rosbag2_storage::BagMetadata & metadata_info)
```

MichaelOrlov commented on rosbag2_py/src/rosbag2_py/_info.cpp:
```diff
@@ -41,12 +42,30 @@ class Info
+    rosbag2_storage::BagMetadata metadata_info)
   {
-    auto metadata_info = read_metadata(uri, storage_id);
+    for (auto topic_info : metadata_info.topics_with_message_count) {
```
```suggestion
    for (const auto & topic_info : metadata_info.topics_with_message_count) {
```

MichaelOrlov commented on rosbag2_py/src/rosbag2_py/_info.cpp:
```diff
@@ -41,12 +42,30 @@ class Info
+    }
+  }
 
+  void print_output_verbose(
+    const std::string & uri,
+    rosbag2_storage::BagMetadata metadata_info)
```
```suggestion
  void print_output_verbose(
    const std::string & uri, const rosbag2_storage::BagMetadata & metadata_info)
```

MichaelOrlov commented on rosbag2_py/src/rosbag2_py/_info.cpp:
```diff
@@ -60,9 +79,19 @@ class Info
     }
 
+    std::unordered_map<std::string, uint64_t> messages_size = {};
+    for (auto & file_info : metadata_info.files) {
```
```suggestion
    for (const auto & file_info : metadata_info.files) {
```

MichaelOrlov reviewed (commented):
@nicolaloi Thanks for addressing the review comments.  
Overall, it looks good among a few nitpicks, mainly related to using const references.

nicolaloi committed:
Apply suggestions from code review

Co-authored-by: Michael Orlov <morlovmr@gmail.com>
Signed-off-by: Nicola Loi <nicolaloi@outlook.com>

MichaelOrlov reviewed (approved):
Thanks looks good to me if CI will pass green.

MichaelOrlov commented:
@nicolaloi The Rpr job failed because need to make rebase.
Please rebase your branch on top of latest `rolling` branch.

nicolaloi committed:
Update timestamp check for new ros bag info test

Signed-off-by: Nicola Loi <nicolaloi@outlook.com>

nicolaloi commented:
@MichaelOrlov I have rebased the branch and I have updated the new size contribution test I have added to match the latest changes in rolling, but now the Rpr test is failing due to an unrelated rcpputils::fs issue, as in the latest rolling merged PR [#1740](https://github.com/ros2/rosbag2/pull/1740) [(log L32773)](https://github.com/ros2/rosbag2/actions/runs/10010496093/job/27671805087#step:3:32773)

MichaelOrlov commented:
@nicolaloi No worries. Those test failures are unrelated to the changes from this PR and test failing because the CI job from GitHub actions uses binaries for the other dependent packages from the latest release on rolling.  
The test failure will go away when we will release a new version of the `rcpputils` package and do rolling synch.

Anyway, I will run CI jobs on the ROS 2 build farm, which compiles all packages from sources. If everything is correct, all tests should pass.

MichaelOrlov commented:
Pulls: ros2/rosbag2#1726
Gist: https://gist.githubusercontent.com/MichaelOrlov/5179a9d2b5bc8685c059758e5d9397de/raw/af0dc6b9cdc1954a535c28a5eeb3e5f43c031848/ros2.repos
BUILD args:  --packages-above-and-dependencies ros2bag rosbag2_cpp rosbag2_py rosbag2_tests
TEST args:  --packages-above ros2bag rosbag2_cpp rosbag2_py rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14281
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=21530)](http://ci.ros2.org/job/ci_linux/21530/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=15810)](http://ci.ros2.org/job/ci_linux-aarch64/15810/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=1025)](http://ci.ros2.org/job/ci_linux-rhel/1025/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=22244)](http://ci.ros2.org/job/ci_windows/22244/)

MichaelOrlov commented:
https://github.com/Mergifyio backport jazzy

jdcast commented:
Will this be backported to Humble?

MichaelOrlov commented:
It is difficult to backport since the code base between Jazzy and Humble is significantly different.  
The major difference is the feature for recording and playback services. 
I have tried to create an automatic backport to humble, but there are a bunch of conflicts and missing files in delta.  Please see comment https://github.com/ros2/rosbag2/pull/1899#issuecomment-2603463783.
I personally don't have the capacity and resources to address all those merge conflicts and make all adjustments.  

If someone will be able to pick up the draft PR https://github.com/ros2/rosbag2/pull/1899 and finalize it by adopting it to humble keeping API/ABI compatibility will be nice.

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-kilted-kaiju-release/43902/1
